api/v1alpha1/common_validation.go

```diff
@@ -0,0 +1,101 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/robfig/cron/v3"
+
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+const (
+	// ValidateSchedulerError defines the error message for ValidateScheduler
+	ValidateSchedulerError = "schedule and duration should be omitted or defined at the same time"
+
+	// ValidatePodchaosSchedulerError defines the error message for ValidateScheduler of Podchaos
+	ValidatePodchaosSchedulerError = "schedule should be omitted"
+
+	// ValidateValueParseError defines the error message for value parse error
+	ValidateValueParseError = "parse value field error:%s"
+)
+
+// ValidateScheduler validates the scheduler and duration
+func ValidateScheduler(duration *string, scheduler *SchedulerSpec, spec *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	schedulerField := spec.Child("scheduler")
+	durationField := spec.Child("duration")
+
+	if duration != nil && scheduler != nil {
+		_, err := cron.ParseStandard(scheduler.Cron)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(schedulerField.Child("cron"),
+				scheduler.Cron,
+				fmt.Sprintf("parse cron field error:%s", err)))
+		}
+
+		_, err = time.ParseDuration(*duration)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(durationField,
+				*duration,
+				fmt.Sprintf("parse duration field error:%s", err)))
+		}
+
+		if len(allErrs) > 0 {
+			return allErrs
+		}
+
+		return nil
+	} else if duration == nil && scheduler == nil {
+		return nil
+	}
+
+	allErrs = append(allErrs, field.Invalid(schedulerField, scheduler, ValidateSchedulerError))
+	return allErrs
+}
+
+// ValidatePodMode validates the value with podmode
+func ValidatePodMode(value string, mode PodMode, valueField *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	switch mode {
+	case FixedPodMode:
+		num, err := strconv.Atoi(value)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(valueField, value,
+				fmt.Sprintf(ValidateValueParseError, err)))
+		}
+
+		if num <= 0 {
+			allErrs = append(allErrs, field.Invalid(valueField, value,
+				fmt.Sprintf("value must be greater than 0 with mode:%s", FixedPodMode)))
+		}
+
+	case RandomMaxPercentPodMode, FixedPercentPodMode:
+		percentage, err := strconv.Atoi(value)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(valueField, value,
+				fmt.Sprintf(ValidateValueParseError, err)))
+		}
+
+		if percentage <= 0 || percentage > 100 {
+			allErrs = append(allErrs, field.Invalid(valueField, value,
+				fmt.Sprintf("value of %d is invalid, Must be (0,100] with mode:%s",
+					percentage, mode)))
+		}
+	}
```

Comment:
Should we check unknown mode here?

Reply:
I think we don't need to check unknown mode. Because we've already used the CRD-Validator to do the same things.

```
// +kubebuilder:validation:Enum=one;all;fixed;fixed-percent;random-max-percent
Mode PodMode `json:"mode"`
```

The unknown mode will occur an error.

Reply:
Great 👍